initial run at creating output tests

gregsdennis opened the pull request:
Begins to address #247 

This is a go at creating some output tests.  It includes 2019-09/2020-12 and draft-next tests for:

- annotations shouldn't be reported for failed validations
- `type` creates a proper node, and _may_ include a message if it fails
- `readOnly` generates an annotation of its value

Hat's off to @karenetheridge who proposed using a schema to validate the output.  This creates a really nice way to target the bit of the output that a test is focused on without requiring an explicit error message or being reliant on specific output unit sequencing.

I still have structural tests planned, but I figured this would be enough to start.

I've also included a README for the folder.

Julian commented on output-tests/content/draft-next/general.json:
```diff
@@ -0,0 +1,43 @@
+            }
+        ]
+    }
+]
```
Can you configure your editor to not strip the newline at the end?

gregsdennis replied:
Might be available in .editorconfig if you don't mind having that.

Julian replied:
Yeah definitely fine with adding one of those.

gregsdennis replied:
✔️

Julian commented on output-tests/content/draft2019-09/general.json:
```diff
@@ -0,0 +1,34 @@
+    {
+        "description": "failed validation produces no annotations",
+        "schema": {
+            "$schema": "https://json-schema.org/draft/2020-12/schema",
```
This is supposed to be 2019 right?

Julian commented on output-tests/README.md:
```diff
@@ -0,0 +1,26 @@
+
+The tests are organized into two categories: content and structure.
+
+Content tests verify that the keywords are producing the correct annotations.  Since there are no requirements on the content of error messages, there's not much that can be verified for them, if they're even generated.  These tests need to extensively cover the annotation behaviors of each keyword.  The expected output format for these tests is `basic`.
```
Not much sense bikeshedding at this point so feel free to ignore, but should we consider using the "existing terminology" here and call these `annotation tests`? That's what they seem to cover, right?

gregsdennis replied:
Maybe I need to return this.  There _are_ some tests that make sure that an error is the right type if present.  I'm not sure if that's useful, but we can discuss.

Julian commented on output-tests/content/draft-next/general.json:
```diff
@@ -0,0 +1,43 @@
+[
```
Again ignore if needed but did you consider inverting the directory structure to `output-tests/draft-next/{content,structure}` rather than `output-tests/{content,structure}/draft-next`? They seem equally easy for implementers to run, but I think the former is mildly easier for us to maintain (requires remembering to `cp -R` only one additional directory when a new draft is released)

Which on the latter reminds me, can you add a step to #571's description when we merge this to remember to include updating these too.

gregsdennis replied:
I can do that.

gregsdennis replied:
Does that issue just remain perpetually open?

Julian replied:
No it was/is meant to go in some sort of "documentation site" for the suite (rather than continuing to pile into the README which was meant more for implementers), but the former is still waiting for the `json-schema.org` rewrite/merge...

Obviously up for ideas if you have em on where to put it instead (including in the README if you think we should "give up" for now...)

gregsdennis replied:
updated

Julian commented on output-tests/README.md:
```diff
@@ -0,0 +1,26 @@
+These tests are intended to validate that implementations are correctly generating output in accordance with the specification.
+
+Output was initially specified with draft 2019-09.  It remained unchanged for draft 2020-12 <!-- do we need explicit tests for 2002-12? -->, but will receive an update with the next release.
```
On explicit tests for 2020-12, personally I think ideally yes but it's fine not to have them initially, otherwise someone with an implementation only supporting 2020-12 (which we even encourage these days?) has no way of running these.

But doesn't seem top priority relative to just getting some tests in.

gregsdennis replied:
Okay. It's just going to be an exact copy of 2019.  Makes maintaining the tests more difficult.

Julian replied:
Could always just symlink the directory too.

Julian replied:
Oh, we can't though, since the `$schema`s will be different, so not precisely the same...

gregsdennis replied:
Ugh.  Fair point.

Julian commented on output-tests/README.md:
```diff
@@ -0,0 +1,26 @@
+
+## Test Files
+
+The content of a test file is the same as the validation tests in `tests/`, however an `output` property has been added to each test case.
```
Probably a decent idea to regardless copy the `test-schema.json` file here and make sure it matches these (and add a test to CI that tests are valid under it in this directory -- which I think should just be some copy pasting).

gregsdennis replied:
I don't know if we need to copy it if it's in the same repo.

One file that I _would_ like to copy is the output schema (that's now updated to the `draft-next` version).  I think referencing that would reduce a lot of boilerplate.

Julian reviewed (commented):
Cool! Looks great overall to me I think. Left some pretty minor comments I think, and may take a second read after your response(s) but yeah think this is good!

jdesrosiers commented on .editorconfig:
```diff
@@ -0,0 +1 @@
+insert_final_newline = true
```
:rofl: Irony :rofl:

gregsdennis replied:
Yeah, my editor doesn't seem to be actually doing it... I'll need to go through the files I changed and update them.

jdesrosiers commented on output-tests/README.md:
```diff
@@ -0,0 +1,30 @@
+These tests are intended to validate that implementations are correctly generating output in accordance with the specification.
+
+Output was initially specified with draft 2019-09.  It remained unchanged for draft 2020-12, but will receive an update with the next release.
```
Didn't we change "instanceLocation" from a URL to jsut a JSON Pointer in 2020-12?

gregsdennis replied:
I thought it was always a pointer; just that I was rendering it as a URI fragment in my implementation (and probably in the spec examples).

jdesrosiers replied:
From 2019-09, https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-02#section-10.3.3

> The location of the JSON value within the instance being validated. The value MUST be expressed as a URI fragment-encoded JSON Pointer.

From 2020-12, https://json-schema.org/draft/2020-12/json-schema-core.html#section-12.3.3

> The location of the JSON value within the instance being validated. The value MUST be expressed as a JSON Pointer.

gregsdennis replied:
Maybe "It remained _largely_ unchanged..."?

jdesrosiers replied:
That works for me

jdesrosiers commented on test-schema.json:
```diff
@@ -33,6 +37,7 @@
 
     "$defs": {
         "test": {
+            "$dynamicAnchor": "foo",
```
I'm sure we can come up with an anchor name other than "foo". Perhaps "test"?

jdesrosiers reviewed (commented):
I haven't tied it out yet, but the concept looks sound.

Julian commented on bin/jsonschema_suite:
```diff
@@ -88,12 +90,17 @@ class SanityTests(unittest.TestCase):
         print(f"Found {len(cls.test_files)} test files")
 
+        cls.output_test_files = list(collect(OUTPUT_ROOT_DIR))
+        assert cls.remote_files, "Didn't find the output test files!"
```
You have a typo here, which is the (first, though maybe not only) reason CI fails. You want

```suggestion
        assert cls.output_test_files, "Didn't find the output test files!"
```

Julian commented on bin/jsonschema_suite:
```diff
@@ -88,12 +90,17 @@ class SanityTests(unittest.TestCase):
 
+        cls.output_test_files = list(collect(OUTPUT_ROOT_DIR))
+        assert cls.remote_files, "Didn't find the output test files!"
+        print(f"Found {len(cls.remote_files)} output test files")
```
And here:
```suggestion
        print(f"Found {len(cls.output_test_files)} output test files")
```

gregsdennis commented:
I've added `$id`s to all of the output validation schemas.  This provides a base URI for the `$ref` to use.

I also implemented this new suite in my lib (PR linked ☝️) and I'm happy to report that I fail the 2019 & 2020 cases! 🤦

gregsdennis commented:
I now pass most of the 2019/2020 cases.  There's one test in each that I can't pass because my approach for `draft-next` doesn't always track the keywords in passing subschemas, so I can't correctly report the full keyword locations.  It's fine though: I just mark those tests as inconclusive as I've just decided I'm not going to fully support those drafts.

jdesrosiers commented:
I'm fine with merging as a starting point. Forgive me for not re-reading the readme changes, but does it say somewhere that people shouldn't rely on these test yet because they are still being developed and may contain errors?

Julian commented:
Probably a very good idea especially considering the last conversation.
